server/test/socketManager-test.js

```diff
@@ -0,0 +1,517 @@
+      let socket;
+      try {
+        socket = bsock.connect(ports.manager);
+        socket.on('connect', async function() {
+          // if we reach this done then the test has passed
+          await socket.destroy();
+          done();
+        });
+
+        socket.on('error', e => {
+          done(new Error(`Could not connect to socket server ${e.message}`));
+        });
+      } catch (e) {
+        done(new Error(`There was an error connecting to the socket: ${e}`));
+      }
+    });
+
+    it('should handle auth', async function() {
+      assert(
+        !socketManager.noAuth,
+        'noAuth should be set to false in the test'
+      );
+
+      const sockets = socketManager.io.sockets.values();
+      for (let socket of sockets) {
+        assert(!socket.channel('auth'), 'Expected no authenticated sockets');
+      }
+
+      const authSocket = () => client.call('auth', apiKey);
+
+      const resp = await authSocket();
+      assert.isNull(
+        resp,
+        'Expected a null response on successful authentication'
+      );
+
+      try {
+        await authSocket();
+        throw 'Expected second authentication to throw an error';
+      } catch (e) {
+        assert(e instanceof Error, e);
+        assert(
+          e.message.match(/Already authed/),
+          'Should include "Already authed" error message'
+        );
+      }
+    });
+  });
+
+  describe('addClients, removeClients, getIdFromPath', function() {
+    let clients, id, pathname;
+    beforeEach('add clients to socketManager', async function() {
+      if (wclient.opened) await wclient.close();
+      if (nclient.opened) await nclient.close();
+      clients = {
+        wallet: wclient,
+        node: nclient
+      };
+      id = 'test123';
+      pathname = `/${id}/`;
+      await socketManager.addClients(id, clients);
+    });
+
+    afterEach('cleanup socketManager clients', async function() {
+      if (socketManager.clients.has(id)) await socketManager.removeClients(id);
+    });
+
+    it('should be able to add new clients keyed to id and type', function() {
+      assert(
+        socketManager.clients.has(id),
+        `Could not find clients under the id ${id}`
+      );
+      const managerClients = socketManager.clients.get(id);
+      assert.instanceOf(
+        managerClients['wallet'],
+        WalletClient,
+        'Expected a wallet client'
+      );
+      assert.instanceOf(
+        managerClients['node'],
+        NodeClient,
+        'Expected a node client'
+      );
+    });
+
+    it('should not allow duplicate clients', async function() {
+      try {
+        await socketManager.addClients(id, clients);
+        assert(false, 'Expected addClients to throw on duplicate');
+      } catch (e) {
+        assert.instanceOf(
+          e,
+          Error,
+          'Expected an error to be thrown on duplicate'
+        );
+        assert(e.message.match(/already exists/), e.message);
```

Comment:
Testing for the exact error is what we want, but testing for a matching error message couples the test to the message text, which should be able to change without breaking tests in my opinion. This isn't an easy problem, but I've seen error codes used before. This is definitely out of the scope of this PR and we may not even need that functionality in this app as I don't think that things will change all too much once they are standardized

Reply:
Yeah, totally agree. I was trying to avoid this but I also wanted to make sure it failed for the right reason which is why in these tests I was using regex. Kind of figured that the error message no matter what should be indicating a problem of "already exist"ing. Totally agree with the hesitancy though.